### ISSUE TITLE
add another class to my YOLOv8 model

### ISSUE THREAD

**Issue by MohadeseKarimi77** [ultralytics/ultralytics#5126]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello. I want to add another class to my YOLOv8 model, making it a total of 81 classes. However, I don't want to use the COCO dataset or any other dataset to preserve those 80 classes because the volume of data for the new class, which is class 81, is substantial. Is there a way I can train the YOLOv8 model for the new class while preserving the existing 80 classes?

**Comment by Kayzwer** [comment]
no, it is impossible, but you can retrain by using the pretrained model, and it will transfer all the weights but not the last layer.

**Comment by glenn-jocher** [comment]
@MohadeseKarimi77 hello,

Thank you for your question. The addition of a new class in YOLOv8 implies that the output dimensionality from the final layer of your model would change, requiring a retraining of your model.

However, you can utilize the concept of **Transfer Learning**. This allows you to leverage the learned features of your current 80-class YOLOv8 model and apply it to your new 81-class problem.

Transfer learning generally implies freezing the weights of the earlier layers, preserving the feature extraction capabilities of the model while only re-training the later part of the model. In this case, you'd want to re-train your model on the new dataset (which includes your additional class) while not freezing the weights of the entire model.

Your original 80 classes won't be 'preserved' exactly in the sense they won't be exactly the same as before, as adding new data to the mix and retraining will likely shift the learned representation slightly. But it's about as close as you can get to 'preserving' the original capability of the model with the addition of a new class.
      
Keep in mind to update the `nc` parameter in your `.yaml` configuration file to reflect the new number of classes, which is 81 in your case.

I hope this helps. If you have additional questions, feel free to ask.
  
Best regards,
Glenn from Ultralytics

**Comment by MohadeseKarimi77** [comment]
Thank you very much. How can I use Transfer Learning in YOLOv8? Another question I have is, approximately how many training and validation data points does each class in the COCO dataset have?

**Comment by glenn-jocher** [comment]
@MohadeseKarimi77 hello,

To apply Transfer Learning in YOLOv8, here's a verbal outline of the steps:

1. Load your existing trained weights: When you initialize your YOLOv8 object for training, make sure to specify your previously trained weights. These weights include the learned features from the 80-class model.
   
2. Configure training for new classes: Update the configuration file (usually a .yaml file) that you're using to specify the number of classes (`nc`) to 81, and provide the names of these classes in a `.names` file.

3. Run the training: When training, YOLOv8 will start with the loaded weights and will then commence the training process on your new dataset.

As for the COCO dataset, it's a bit challenging to provide an exact number of training and validation datapoints for each class. This is because the COCO dataset is instance-based, not class-based. It means that each image can contain multiple instances of different classes. 

However, if you are looking at 'instances per category' in the COCO dataset, you can find such statistics in the COCO dataset paper. There, they discuss the total number of instances per category in the dataset. It's an average number, so please be aware this doesn't mean each category or class has the exact same number of instances.

Hope this answers your questions. Please, don't hesitate to ask if you need more clarification.

Best Regards,
Glenn from Ultralytics

**Comment by MohadeseKarimi77** [comment]
I have another question. When training the model, in addition to having a dataset for the new class, do I also need to provide a dataset for the previous 80 classes, correct? And because the features have already been extracted, does the model maintain high accuracy for the previous 80 classes with a small-sized dataset?

**Comment by glenn-jocher** [comment]
@MohadeseKarimi77 hello,

Thank you for your question. When you're adding a new class and leveraging Transfer Learning, you are correct that you would typically have data for all the classes - in your case, the 80 previous classes and the new one.

However, the power of Transfer Learning lies in its ability to use previously learned features from a different but related task to improve learning in a new task. So, theoretically, if you decided not to provide data for the previous 80 classes during retraining, the model would still have a general understanding of those classes from previous training.

That being said, if you don't expose your model to the previous classes during the retraining phase, there's a chance that its performance on those classes could degrade over time. Even though the features have been extracted, it's beneficial for the model to continue learning from those classes to maintain its effectiveness.

To ensure a high accuracy for the previous 80 classes while focusing on the new class, consider using a balanced dataset that includes examples from all 81 classes. Naturally, the more data you provide to your model, the better it tends to perform.

I hope this helps. If you have further questions, please feel free to ask.

Best,
Glenn from Ultralytics

**Comment by MohadeseKarimi77** [comment]
Thank you very much

**Comment by glenn-jocher** [comment]
@MohadeseKarimi77 hello,

It appears that your message might have been cut in the process, as there's no GitHub issue mentioned.

Please provide the issue or question again so I can provide the best possible assistance in relation to YOLOv8. Feel free to ask anything you wish about the YOLOv8 framework— from training, validation, prediction, tracking, benchmarking, or any technical specifications or enhancements.

Looking forward to your response.

Best Regards,
Glenn from Ultralytics

**Comment by MohadeseKarimi77** [comment]
No, thank you very much. I don't have any more questions for now. If I have another question in the future, I'll ask you.

**Comment by MohadeseKarimi77** [comment]
Do I need a dataset as large as COCO for the previous 80 classes, or can I train on a smaller dataset and still maintain an accuracy of over 90% for those 80 classes? Since the COCO dataset is very large and training the model on it is time-consuming and challenging. Additionally, I would need a substantial dataset for new classes as well, which would make the dataset quite extensive overall.

**Comment by MohadeseKarimi77** [comment]
And I have another question.
You previously provided me with this response: "However, the power of Transfer Learning lies in its ability to use previously learned features from a different but related task to improve learning in a new task. So, theoretically, if you decided not to provide data for the previous 80 classes during retraining, the model would still have a general understanding of those classes from previous training."
In theory, it is as you described, but practically, it is not possible, is it? Because I trained the YOLOv8 model only with the dataset related to the new class, but after completing the training and making predictions, the model couldn't recognize those previous 80 classes and could only recognize the new class.

**Comment by glenn-jocher** [comment]
@MohadeseKarimi77 

Hello!

When using transfer learning, you're correct that while the model should theoretically retain some feature learning from prior classes, in practical terms, if a model isn't exposed to a particular class during training, its performance on identifying that class can degrade. It's similar to a 'use-it-or-lose-it' situation. If you're training your model only on the new class, it's effectively learning to identify only that class and can potentially 'forget' the others.

So, if the 81st class is mixed with examples of the previous 80 classes, it's likely to perform better when identifying those 80 classes as compared to the scenario where it's trained with only the 81st class. Therefore, it would be advisable to add examples of the previously known classes when retraining your model with the new class.

With regards to dataset size, it's indeed true that the larger the dataset, the better the generalization of the model. However, you necessarily don't need a dataset as large as COCO for the previous 80 classes to achieve high accuracy. You can train on a smaller, balanced dataset across all 81 classes and still achieve good accuracy. However, the dataset should be representative of the problem space, and large enough to contain sufficient variability in the images that the model will encounter.

I hope this helps! If you have further questions, please feel free to ask. 

Best,
Glenn from Ultralytics

**Comment by MohadeseKarimi77** [comment]
I got it. Thank you very much for your guidance.🙂

**Comment by glenn-jocher** [comment]
@MohadeseKarimi77 

Hello!

You're welcome! I'm glad I could assist you. If you have any more questions in the future about using YOLOv8 for your object detection tasks, feel free to ask. Your questions are always welcome!

Happy coding, and best of luck with your project!

Best Regards,
Glenn

**Comment by ParimiHarsha** [comment]
@MohadeseKarimi77 I am trying to solve a similar problem as you. Did the solution given by @glenn-jocher help you out? Please let me know how you solved this problem.

Thanks!!

**Comment by glenn-jocher** [comment]
Hello @ParimiHarsha,

Yes, Glenn's advice was really helpful! I incorporated examples of my new class along with some sample data from the previous classes during retraining. This approach allowed the model to remember the old classes while learning the new one. Just make sure your training data is varied enough to cover the different scenarios your model might face. 

Here's a simple snippet showing how to set up training with YOLOv8:

```bash
yolo train data=combined_dataset.yaml model=yolov8n.pt epochs=100
```

Ensure `combined_dataset.yaml` includes paths for both old and new class images. Hope this helps! 😊

**Comment by BhavyaDedhia** [comment]
> I got it. Thank you very much for your guidance.🙂

Hello,
I am also trying to add classes to a YOLO model. Have you trained it and tried it? If so, could you please help me with it? I attempted to add new classes to the model, but I am not getting any results. The model is not even detecting the original 80 classes. If you have successfully done this, please guide me on how you achieved it.

**Comment by germancruzram** [comment]
> Hola@ParimiHarsha,
> 
> Sí, el consejo de Glenn fue realmente útil. Incorporé ejemplos de mi nueva clase junto con algunos datos de muestra de las clases anteriores durante el reentrenamiento. Este enfoque le permitió al modelo recordar las clases anteriores mientras aprendía la nueva. Solo asegúrese de que sus datos de entrenamiento sean lo suficientemente variados para cubrir los diferentes escenarios que su modelo podría enfrentar.
> 
> A continuación se muestra un fragmento simple que muestra cómo configurar el entrenamiento con YOLOv8:
> 
> ```shell
> yolo train data=combined_dataset.yaml model=yolov8n.pt epochs=100
> ```
> 
> Asegúrese `combined_dataset.yaml`de incluir rutas para las imágenes de clase antiguas y nuevas. ¡Espero que esto te ayude! 😊

Interesting topic, I would like to inquire about the case where my pre-trained model (x classes) was based on YOLOv8 Small. I observe that it always downloads the YOLOv8n.pt model by default. How can I control the transfer learning to be performed on my YOLOv8s model?

**Comment by pderrenger** [comment]
You can definitely control the transfer learning to be performed on your YOLOv8s model. Instead of using the default YOLOv8n.pt model, specify your pre-trained YOLOv8s model in the training command. This way, the training will continue from your YOLOv8s model rather than starting from scratch with YOLOv8n.pt. Just replace `yolov8n.pt` with the path to your YOLOv8s model. This should help you maintain the performance of your pre-trained classes while adding new ones.

**Comment by Dimasilham7** [comment]
> You can definitely control the transfer learning to be performed on your YOLOv8s model. Instead of using the default YOLOv8n.pt model, specify your pre-trained YOLOv8s model in the training command. This way, the training will continue from your YOLOv8s model rather than starting from scratch with YOLOv8n.pt. Just replace `yolov8n.pt` with the path to your YOLOv8s model. This should help you maintain the performance of your pre-trained classes while adding new ones.

Hey, I am doing transfer learning on my process. The result should be that my new model added a new class, but the result I got is that the class on my transfer learning model is being replaced. Doesn't it should be added and not replaced?

**Comment by olfy002** [comment]
Is there any documentation or video for how to add a new class to the pretrained COCO model? This would be very helpfull!

**Comment by Dimasilham7** [comment]
> Is there any documentation or video about how to train the pretrained COCO model with new classes and to use the 80 COCO classes + new classes? This would be very helpfull!

I think you can check on [here](https://y-t-g.github.io/tutorials/yolov8n-add-classes/), this should be help your questions

**Comment by glenn-jocher** [comment]
For official guidance on training a pretrained COCO model with new classes while retaining the original 80, please refer to the [Ultralytics YOLO documentation](https://docs.ultralytics.com).
